contracts/periphery/mintburn/HyperCoreFlowExecutor.sol

```diff
@@ -0,0 +1,1195 @@
+                finalCoreTokenInfo.coreIndex,
+                totalEVMAmountToSponsor,
+                finalCoreTokenInfo.tokenInfo.evmExtraWeiDecimals
+            );
+            cumulativeSponsoredAmount[finalToken] += totalEVMAmountToSponsor;
+        }
+
+        uint128 cloid = ++nextCloid;
+        swapHandler.submitLimitOrder(finalTokenInfo, limitPriceX1e8, sizeX1e8, cloid);
+
+        pendingSwaps[quoteNonce] = PendingSwap({
+            finalRecipient: finalRecipient,
+            finalToken: finalToken,
+            minCoreAmountFromLO: guaranteedLOOut,
+            sponsoredCoreAmountPreFunded: totalCoreAmountToSponsor,
+            limitOrderCloid: cloid
+        });
+        pendingQueue[finalToken].push(quoteNonce);
+        cloidToQuoteNonce[cloid] = quoteNonce;
+
+        emit SwapFlowInitialized(
+            quoteNonce,
+            finalRecipient,
+            finalToken,
+            amountInEVM,
+            totalEVMAmountToSponsor,
+            finalCoreSendAmount,
+            finalTokenInfo.assetIndex,
+            cloid
+        );
+    }
+
+    /// @notice Finalizes pending queue of swaps for `finalToken` if a corresponsing SwapHandler has enough balance
+    function finalizePendingSwaps(address finalToken, uint256 maxToProcess) external {
+        FinalTokenInfo memory finalTokenInfo = _getExistingFinalTokenInfo(finalToken);
+        CoreTokenInfo memory coreTokenInfo = coreTokenInfos[finalToken];
+
+        require(lastPullFundsBlock[finalToken] < block.number, "Can't finalize twice in the same block");
+
+        uint256 head = pendingQueueHead[finalToken];
+        bytes32[] storage queue = pendingQueue[finalToken];
+        if (head >= queue.length || maxToProcess == 0) return;
+
+        // Note: `availableCore` is the SwapHandler's Core balance for `finalToken`, which monotonically increases
+        uint64 availableCore = HyperCoreLib.spotBalance(address(finalTokenInfo.swapHandler), coreTokenInfo.coreIndex);
+        uint256 processed = 0;
+
+        while (head < queue.length && processed < maxToProcess) {
+            bytes32 nonce = queue[head];
+
+            PendingSwap storage pendingSwap = pendingSwaps[nonce];
+            uint64 totalAmountToForwardToUser = pendingSwap.minCoreAmountFromLO +
+                pendingSwap.sponsoredCoreAmountPreFunded;
+            if (availableCore < totalAmountToForwardToUser) {
+                break;
+            }
+
+            finalTokenInfo.swapHandler.transferFundsToUserOnCore(
+                finalTokenInfo.assetIndex,
+                pendingSwap.finalRecipient,
+                totalAmountToForwardToUser
+            );
+
+            emit SwapFlowCompleted(
+                nonce,
+                pendingSwap.finalRecipient,
+                pendingSwap.finalToken,
+                pendingSwap.sponsoredCoreAmountPreFunded,
+                totalAmountToForwardToUser
+            );
+
+            availableCore -= totalAmountToForwardToUser;
```

Comment:
Isn't there reentrancy risk from the `transferFundsToUserOnCore` call() that appears above these internal updates to vars?

Reply:
For this specific case, I don't think so.

The CoreWriter is a trusted precompile contract and only emits an event (no call-outs). Still may make sense to have re-entrancy guards, though. I'm thinking through that now.